keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,279 @@
+1. The Sidecar deletes the OB’s secret.
+1. The central controller deletes the secret’s copy in the app’s namespace.
+1. The central controller removes the finalizers from the OBC, OB, and Secret.
+1. The OBC and its secret are garbage collected and the OB remains.
+
+##  Custom Resource Definitions
+
+#### ObjectBucketClaim (OBC)
+
+A user facing API object representing a request for a bucket, or access to an existing bucket. OBCs are created by users in their namespaces. Once the request is fulfilled, the OBC is “bound” to an Object Bucket (OB). The binding is used to mark the request as fulfilled and prevent further binds to the OB.
+
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: ObjectBucketClaim
+metadata:
+  name: [1]
+  namespace:
+  labels:
+    “cosi.io/driver”: [2]
+  finalizers:
+  - cosi.io/finalizer [3]
+spec:
+  bucketName: [4]
+  generateBucketName: [5]
+  bucketClassName: [6]
+  objectBucketName: [7]
+  secretName: [8]
+  driverSecretName: [9]
+status:
+  phase:
+  conditions: []ObjectBucketClaimCondition
+```
+1. `name`: the metadata name of the OBC; however this name can be generated and thus cannot be relied on to predictably name other related resources, eg. secrets. 
+1. `labels`: central controller adds the label to its managed resources for easy GET ops.  Value is the driver name returned by GetDriverInfo() rpc*.
+1. `finalizers`: central controller adds the finalizer to defer OBC deletion until backend deletion ops succeed.
+1. `bucketName`: Desired name of the bucket to be created**.
+1. `generateBucketName`: Prefix to a randomly generated name. Mutually exclusive with `bucketName`**.
+1. `bucketClassName`: Name of the target BucketClass**.
+1. `objectBucketName`: Name of a bound OB.
+   - Injected by the central controller during greenfield ops.
+   - Defined by the OBC author for static and dynamic brownfield ops.
+1. `secretName`: Desired name of the app's secret for greenfield. Fails if secret exists. Defined here so that app deployment (where the secret name is required) is independent of OBC creation.
+1. `driverSecretName`: Name of the driver's secret with the namespace assumed to be "cosi-system".
+   - Injected by the central controller during greenfield and dynamic brownfield ops.
+   - Defined by the OBC author for static brownfield ops.
+
+\* Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
+
+** Ignored for brownfield usage.
+
+#### ObjectBucket (OB)
+A cluster-wide scoped resource representing the bucket. The OB is expected to store stateful data relevant to bucket deprovisioning. The OB is bound to the OBC in a 1-1 mapping.
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: ObjectBucket
+Metadata:
+  name: [1]
+  namespace: [2]
+  labels:
+    “cosi.io/driver”: [3]
+  finalizers:
+  - "cosi.io/finalizer" [4]
+spec:
```

Comment:
Add: the `Bucket` contains all of the relevant info from the BucketClass at the time the Bucket is created.

Reply:
@jsafrane I _think_  the idea of a handle was discussed, and due to the 1:many mapping of Bucket:BucketRequest, rather than a list of BucketRequests (which needs to be maintained), a handle is omitted. This means `kubectl describe bucket` won't show the BucketRequest that generated this Bucket.